near-sdk/src/collections/tree_map.rs

```diff
@@ -0,0 +1,1772 @@
+use std::ops::Bound;
+use borsh::{BorshDeserialize, BorshSerialize};
+
+use crate::collections::{append, next_trie_id, Vector};
+use crate::collections::UnorderedMap;
+
+/// TreeMap based on AVL-tree
+///
+/// Runtime complexity (worst case):
+/// - `get`/`contains_key`:     O(1) - UnorderedMap lookup
+/// - `insert`/`remove`:        O(log(N))
+/// - `min`/`max`:              O(log(N))
+/// - `above`/`below`:          O(log(N))
+/// - `range` of K elements:    O(Klog(N))
+///
+#[derive(BorshSerialize, BorshDeserialize)]
+pub struct TreeMap<K, V> {
+    root: u64,
+    val: UnorderedMap<K, V>,
+    tree: Vector<Node<K>>,
+}
+
+#[derive(Copy, Clone, BorshSerialize, BorshDeserialize)]
+pub struct Node<K> {
+    id: u64,
+    key: K,             // key stored in a node
+    lft: Option<u64>,   // left link of a node
+    rgt: Option<u64>,   // right link of a node
+    ht: u64,            // height of a subtree at a node
+}
+
+impl<K> Node<K>
+    where
+        K: Ord + Copy + BorshSerialize + BorshDeserialize
+{
+    fn of(id: u64, key: K) -> Self {
+        Self {
+            id,
+            key,
+            lft: None,
+            rgt: None,
+            ht: 1,
+        }
+    }
+}
+
+impl<K, V> Default for TreeMap<K, V>
+    where
+        K: Ord + Copy + BorshSerialize + BorshDeserialize,
+        V: Copy + BorshSerialize + BorshDeserialize,
+{
+    fn default() -> Self {
+        Self::new(next_trie_id())
+    }
+}
+
+
+impl<K, V> TreeMap<K, V>
+    where
+        K: Ord + Copy + BorshSerialize + BorshDeserialize,
+        V: Copy + BorshSerialize + BorshDeserialize,
+{
+    pub fn new(id: Vec<u8>) -> Self {
+        Self {
+            root: 0,
+            val: UnorderedMap::new(append(&id, b'v')),
+            tree: Vector::new(append(&id, b'n')),
+        }
+    }
+
+    pub fn len(&self) -> u64 {
+        self.tree.len() as u64
+    }
+
+    pub fn clear(&mut self) {
+        self.root = 0;
+        self.val.clear();
+        self.tree.clear();
+    }
+
+    fn node(&self, id: u64) -> Option<Node<K>> {
+        self.tree.get(id)
+    }
+
+    fn save(&mut self, node: &Node<K>) {
+        if node.id < self.len() {
+            self.tree.replace(node.id, node);
+        } else {
+            self.tree.push(node);
+        }
+    }
+
+    /// Return height of the tree - number of nodes on the longest path starting from the root node.
+    pub fn height(&self) -> u64 {
+        self.node(self.root).map(|n| n.ht).unwrap_or_default()
+    }
```

Comment:
This should be private I think in case we want to switch from AVL

Reply:
done